js/src/elements/form-behavior.ts

```diff
@@ -0,0 +1,744 @@
+		alertText.id = alertId;
+		alertText.className = 'font-size-md font-weight-semibold margin-bottom-xs display-inline-flex';
+		alertText.innerText = this.locStrings.pleaseFixTheFollowingIssues;
+
+		const errorList = document.createElement('ul');
+		errorList.setAttribute('aria-label', 'Validation errors');
+		errorList.classList.add('margin-left-sm-tablet');
+
+		errorAlert.append(alertText, errorList);
+		formLayout.appendChild(errorAlert);
+
+		return { errorAlert, errorList };
+	}
+
+	getErrorAlert(form: HTMLFormElement) {
+		const errorAlert = form.querySelector<HTMLDivElement>('.form-error-container .alert');
+		if (errorAlert) {
+			return {
+				errorAlert,
+				errorList: errorAlert.lastElementChild as HTMLUListElement
+			};
+		}
+		return this.createErrorAlert(form);
+	}
+
+	validateRequired(input: HTMLValueElement, label: string): string | null {
+		if (input.validity.valueMissing) {
+			return `${this.locStrings.inputRequired.replace(
+				'{inputLabel}',
+				customElements.get(input.localName) ? `A selection for "${label}"` : label
+			)}`;
+		}
+		return null;
+	}
+
+	validateMinLength(input: HTMLValueElement | HTMLInputElement, label: string): string | null {
+		if (
+			(input instanceof HTMLTextAreaElement || input instanceof HTMLInputElement) &&
+			(input.validity.tooShort || (input.minLength > 0 && input.value.length < input.minLength))
+		) {
+			return `${this.locStrings.inputMinLength
+				.replace('{inputLabel}', label)
+				.replace('{minLength}', input.minLength.toString())}`;
+		}
+		return null;
+	}
+
+	validateMaxLength(input: HTMLValueElement, label: string): string | null {
+		if (
+			(input instanceof HTMLTextAreaElement || input instanceof HTMLInputElement) &&
+			(input.validity.tooLong ||
+				(input.maxLength > 0 && input.value.length > input.maxLength) ||
+				input.value.length > 255)
+		) {
+			return `${this.locStrings.inputMaxLength
+				.replace('{inputLabel}', label)
+				.replace('{maxLength}', input.maxLength > 0 ? input.maxLength.toString() : '255')}`;
+		}
+		return null;
+	}
+
+	validateTagSelector(input: HTMLInputElement | HTMLValueElement, label: string): string | null {
+		if (input instanceof HTMLInputElement && input.classList.contains('tag-input')) {
+			const min = input.getAttribute('minTags');
+			const max = input.getAttribute('maxTags');
+			const tagCount = input.value === '' ? 0 : input.value.split(',').length;
+
+			// if no min or max, no need to validate
+			if (!min || !max) {
+				return null;
+			}
+
+			if (!tagCount || tagCount < Number(min) || tagCount > Number(max)) {
+				return `${this.locStrings.youMustSelectBetweenMinAndMaxTags
+					.replace('{min}', min)
+					.replace('{max}', max)
+					.replace('{tagLabel}', label.toLocaleLowerCase())}`;
+			}
+			return null;
+		}
+		return null;
+	}
+
+	async validateForm(
+		form: HTMLFormElement,
+		displayValidity = true,
+		scope: Element = form,
+		customElements?: Element[]
```

Comment:
let's query these at validation time rather than passing them in for more flexibility.

Reply:
Combined with your loop comment below